03b_ipyautoui.ipynb

```diff
@@ -6,24 +6,678 @@
+    "\n",
+    "Here we will implement 2 because it is a little different than what we did in the first round of the dashboard. It also places the responsibility of deciding the correct course of action (increase `window_size` or decrease `polynomial-order`) on the user. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Our approach is to observe the `_value` of the widget, try to make a valid model out of it, and if that fails set the widget to the old value. `ipyautoui` will take care of displaying an appropriate error message for us.\n",
+    "\n",
+    "The observe below is created using a [*closure*](https://en.wikipedia.org/wiki/Closure_(computer_programming)), which is a function that is created inside of another function where the inner function uses some of the variables from the outer function. Here we do that instead of treating `ui` as a global variable."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#| export\n",
+    "\n",
+    "from pydantic import ValidationError"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#| export\n",
+    "\n",
+    "def make_enforcer(ui):\n",
+    "    \"\"\"\n",
+    "    Make a function that can be used to observe changes on a \n",
+    "    user interface element.\n",
+    "\n",
+    "    Parameters\n",
+    "    ----------\n",
+    "\n",
+    "    ui: an AutoUi widget\n",
+    "\n",
+    "    Returns\n",
+    "    -------\n",
+    "\n",
+    "    callable\n",
+    "        A function that can be used as the observer of a traitlets event.\n",
+    "    \"\"\"\n",
+    "    def constraint_enforcer(change):\n",
+    "        \"\"\"\n",
+    "        Reset widget to the most recent valid value if the new\n",
+    "        value results in an invalid value.\n",
+    "        \"\"\"\n",
+    "        try:\n",
+    "            # Every AutoUi widget has a copy of the model class\n",
+    "            # We'll try validating the value in change[\"new\"] and see if it works\n",
+    "            ui.model.model_validate(change[\"new\"])\n",
+    "        except ValidationError:\n",
+    "            # That failed, so reset the ui to the old value\n",
+    "            ui.value = change[\"old\"]\n",
+    "\n",
+    "    return constraint_enforcer"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Run the cell below to add the observer and manipulate the controls to see the effect that the obserer has."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#| export\n",
+    "controls.observe(make_enforcer(controls), \"_value\")"
```

Comment:
Am I supposed to be able to trigger `Value error, Polynomial order must be smaller than window size [type=value_error, input_value={'year_range': (1915, 200..., 'polynomial_order': 8}, input_type=dict]` here still? Because I am able to do it consistently.

I feel like I understood your verbal description, but without digging into it, I don't understand how `ui.model.model_validate(change["new"])` actually corrects the model if not in the way of version 1.

Reply:
I'll address this in a separate PR*, but the key line is in the `except` clause, where the `ui`'s `value` is reset to `change["old"]`.  The `except` is hit only if the new value makes an invalid model. The idea was that you still get the error display but the widget won't allow an invalid value to be set.

*in the interests of getting this in so that you can make the additions about `functools.partial`